pkg/infradb/subscriberframework/actionbus/actionbus.go

```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (c) 2023-2024 Intel Corporation, or its subsidiaries.
+// Copyright (c) 2024 Ericsson AB
+
+// Package actionbus holds implementation for subscribing and receiving actions
+package actionbus
+
+import (
+	"fmt"
+	"log"
+	"sync"
+)
+
+// ABus holds the ActionBus object
+var ABus = NewActionBus()
+
+// ActionBus holds the action bus info
+type ActionBus struct {
+	subscribers    map[string][]*Subscriber
+	actionHandlers map[string]ActionHandler
+	subscriberL    sync.RWMutex
+	mutex          sync.RWMutex
+}
+
+// Subscriber holds the info for each subscriber
+type Subscriber struct {
+	Name string
+	Ch   chan interface{}
+	Quit chan bool
+}
+
+// ActionHandler handles the action requests that arrive
+type ActionHandler interface {
+	HandleAction(string, *ActionData)
+}
+
+// ActionData holds the data for each action
+type ActionData struct {
+	ErrCh chan error
+}
+
+// StartSubscriber will be called by the modules to initialize and start listening for actions
+func (a *ActionBus) StartSubscriber(moduleName, actionType string, actionHandler ActionHandler) {
+	if !a.subscriberExist(actionType, moduleName) {
+		subscriber := a.Subscribe(moduleName, actionType, actionHandler)
+
+		go func() {
+			for {
+				select {
+				case action := <-subscriber.Ch:
+					log.Printf("\nSubscriber %s for %s received \n", moduleName, actionType)
+
+					handlerKey := moduleName + "." + actionType
+					if handler, ok := a.actionHandlers[handlerKey]; ok {
+						if actionData, ok := action.(*ActionData); ok {
+							handler.HandleAction(actionType, actionData)
+						} else {
+							log.Println("error: unexpected action type")
+						}
+					} else {
+						log.Println("error: no action handler found")
+					}
+				case <-subscriber.Quit:
+					close(subscriber.Ch)
+					return
+				}
+			}
+		}()
+	}
+}
+
+// NewActionBus initializes an ActionBus object
+func NewActionBus() *ActionBus {
+	return &ActionBus{
+		subscribers:    make(map[string][]*Subscriber),
+		actionHandlers: make(map[string]ActionHandler),
+	}
+}
+
+// NewActionData initializes an ActionData object
+func NewActionData() *ActionData {
+	return &ActionData{
+		ErrCh: make(chan error),
+	}
+}
+
+// Subscribe api provides registration of a subscriber to the given action
+func (a *ActionBus) Subscribe(moduleName, actionType string, actionHandler ActionHandler) *Subscriber {
+	a.subscriberL.Lock()
+	defer a.subscriberL.Unlock()
+
+	subscriber := &Subscriber{
+		Name: moduleName,
+		Ch:   make(chan interface{}),
+		Quit: make(chan bool),
+	}
+
+	a.subscribers[actionType] = append(a.subscribers[actionType], subscriber)
+	a.actionHandlers[moduleName+"."+actionType] = actionHandler
```

Comment:
if we rely on string concatenation somewhere, we need to make sure that the concatenation is placed into a  dedicated function and used where it is needed. It is too easy to make a typo and get a mistake

Reply:
Do you think it worths to create a new function for just one line of code. Also the concatenation is based on variables not text (except the "." of course) so the golang compiler will obect during the build if there is any typo. WDYT ?

Reply:
ok I will create a function then that does this concatenation

Reply:
Done

Reply:
Maybe `ComposeActionHandlerName` instead of `ConcatenateModuleNameWithType`?

Reply:
`ComposeHandlerName`?

Reply:
Done